Improve Recommended content queryset

sks444 opened the pull request:
#### Description of changes

Recommended section on the articles page is very expensive in terms of queries.._probably on the other page type too_.

With this `/articles/chinas-shadow-banking-sector-pushed-light/` as an example I have managed to get the queries from 123 to 76:

- Use [cache_property](https://docs.djangoproject.com/en/3.1/topics/performance/#cached-property) whenever we have an expensive method getting executed multiple times. E.g. [self.recommneded_content](https://github.com/CIGIHub/cigionline/blob/0cf98cb0fe42587db63cd291380b3d5a1af67399/templates/articles/article_page.html#L146) is getting called two times in the template...so using cache_property on the `recommneded_content` method will minimise the query from..e.g. 123 to 111.
- [Ordering is expensive](https://docs.djangoproject.com/en/3.0/topics/db/optimization/#don-t-order-results-if-you-don-t-care) and shouldn't be done on objects when not needed..specially on the fields which are not indexed. E.g. in this change we have avoided the use of [topics_sorted](https://github.com/CIGIHub/cigionline/blob/0cf98cb0fe42587db63cd291380b3d5a1af67399/core/models.py#L330) method.
- Use specific() on [ContentPage](https://github.com/CIGIHub/cigionline/blob/0cf98cb0fe42587db63cd291380b3d5a1af67399/core/models.py#L399) queryset as well, given that this is an abstract class and original page will be different from this. This will also avoid specific getting executed on each object in the [template](https://github.com/CIGIHub/cigionline/blob/0cf98cb0fe42587db63cd291380b3d5a1af67399/templates/includes/recommended.html#L18). Saving us a lot of queries..e.g. from 109 to 76

gabrieltanchen commented:
@sks444 If ordering without an index is the issue, why wouldn't we create an index on title rather than ordering in the template?

sks444 commented:
Hi @gabrieltanchen, sorry for the late reply on this, haven't been doing well.

Noticed the title field in [TopicPage](https://github.com/CIGIHub/cigionline/blob/5151c43245903040ddfe4f09c6e4c6494e110d10/research/models.py#L242) model is coming from wagtail Page model, which makes it not possible to use the model indexing for title at TopicPage.

I have reverted to use topics_sorted method anyway as the number of items weren't that big there isn't much gain in number of queries here.

Also, following the discussion from [slack](https://cigionline.slack.com/archives/C01B9KVHS2D/p1617805970029500), I have added the logic to only fetch 3 recommended articles added by the editor.
